main.cpp

```diff
@@ -96,224 +98,448 @@ inf
-    // --------
-    std::cout << "Initial value of dt: " << dt.value << std::endl;
-    std::cout << "Initial value of it: " << it.value << std::endl;
-    // --------
-    std::cout << "Use of function concatenation (mixed type arguments) " << std::endl;
-    std::cout << "New value of dt = (dt * it) / 5.0f + ft = " << (dt.multiply(it).divide(5.0f).add(ft).value) << std::endl;
+    DoubleType& add(double rhs);
+    DoubleType& subtract(double rhs);
+    DoubleType& multiply(double rhs);
+    DoubleType& divide(double rhs);
+
+    DoubleType& add(const FloatType& ft);
+    DoubleType& subtract(const FloatType& ft);
+    DoubleType& multiply(const FloatType& ft);
+    DoubleType& divide(const FloatType& ft);
+
+    DoubleType& add(const DoubleType& dt);
+    DoubleType& subtract(const DoubleType& dt);
+    DoubleType& multiply(const DoubleType& dt);
+    DoubleType& divide(const DoubleType& dt);
+
+    DoubleType& add(const IntType& it);
+    DoubleType& subtract(const IntType& it);
+    DoubleType& multiply(const IntType& it);
+    DoubleType& divide(const IntType& it);
+};
 
-    std::cout << "---------------------\n" << std::endl; 
-    
-    // Intercept division by 0
-    // --------
-    std::cout << "Intercept division by 0 " << std::endl;
-    std::cout << "New value of it = it / 0 = " << it.divide(0).value << std::endl;
-    std::cout << "New value of ft = ft / 0 = " << ft.divide(0).value << std::endl;
-    std::cout << "New value of dt = dt / 0 = " << dt.divide(0).value << std::endl;
+struct IntType
+{
+    int* value;
 
-    std::cout << "---------------------\n" << std::endl; 
+    IntType(int iVal) : value(new int(iVal)) {}
+    ~IntType()
+    {
+        delete value;
+        value = nullptr;
+    }
 
-    std::cout << "good to go!\n";
+    IntType& add(int rhs);
+    IntType& subtract(int rhs);
+    IntType& multiply(int rhs);
+    IntType& divide(int rhs);
+
+    IntType& add(const FloatType& ft);
+    IntType& subtract(const FloatType& ft);
+    IntType& multiply(const FloatType& ft);
+    IntType& divide(const FloatType& ft);
+
+    IntType& add(const DoubleType& dt);
+    IntType& subtract(const DoubleType& dt);
+    IntType& multiply(const DoubleType& dt);
+    IntType& divide(const DoubleType& dt);
+
+    IntType& add(const IntType& it);
+    IntType& subtract(const IntType& it);
+    IntType& multiply(const IntType& it);
+    IntType& divide(const IntType& it);    
+};
 
-    return 0;
+//--------
+
+FloatType& FloatType::add(float rhs)
+{
+    *value += rhs;
+    return *this;
 }
 
+FloatType& FloatType::subtract(float rhs)
+{
+    *value -= rhs;
+    return *this;
+}
 
+FloatType& FloatType::multiply(float rhs)
+{
+    *value *= rhs;
+    return *this;
+}
 
-/*
- MAKE SURE YOU ARE NOT ON THE MASTER BRANCH
+FloatType& FloatType::divide(float rhs)
+{
+    if(rhs == 0.0f)
+    {
+        std::cout << "warning: floating point division by zero!" << std::endl;
+    }
+    if(value != nullptr) *value /= rhs;
```

Comment:
Be consistent with your coding style for `if` expressions.
You've got curly braces and the `if` body on its own line in the first `if` expression
but the 2nd one doesn't have curly braces, nor is the `if` body on its own line.
Pick one style or the other
- [ ] revise